### ISSUE TITLE
Need to find solutions to issues with third party cookies and sessions (in certain situations)

### ISSUE THREAD

**Issue by rishabhpoddar** [supertokens/supertokens-core#280]
TLDR;
- Situation 1: Issues with using iframes which queries an API domain that doesn't share the same base domain as the site where the iframe will be embedded in.
- Situation 2: If your API and website domain do not share the same base domain (for example API is on `api.example.com` and website is not on `*.example.com` and not on `example.com`), sessions will not work on Safari.

The solutions to the above and more details on them is below. Please feel free to [join our Discord](https://supertokens.io/discord) to ask questions about this.

# Situation 1

### What is the situation?
- When your frontend needs to run in an iframe that can be embedded in any website & third parties cookies are blocked.
- Some more info about this here: https://github.com/supertokens/supertokens-core/issues/242

### What happens in this situation?
- The browser does not allow sending of cookies in API calls that the iframe makes, nor does it allow the JS in the iframe to read cookies / localstorage.
- This means that sessions with cookies or localstorage won't work and the user can't use the app when they are logged in.

### What is the solution?
- Apps have to ask for permission from the user to allow sharing of data. For example, Safari has a [Storage Access API](https://webkit.org/blog/8124/introducing-storage-access-api/) which apps can use as long as the end user has interacted with the iframe directly.
- Ask users to allow cross site cookies (via browser settings).

# Situation 2

### What is the situation?
If your website and API domain do not share the same top level / base domain (for example API domain is `api.abc.com` or `abc.com` and website domain is `xyz.com`), Safari won't send cookies if third party cookies are blocked.

### What happens in this situation?
- In this case, the `sameSite` value will be `none`, and whilst this would make chrome and firefox work properly, Safari still blocks sending cookies on API calls.
- It does set cookies in the browser though, but just doesn't send them. It seems like this was a bug previously where it treated `sameSite=none` as `sameSite=strict`, but that has been solved, and yet this issue exists.
- This means that APIs will throw an `unauthorised` error, and the user will be logged out.

### What is the solution?
- [Recommended & Easy]: Set up a reverse proxy to the API domain. This reverse proxy will have the same top level domain as the website domain and the website will query this intermediate domain instead. This will allow the cookies to have `sameSite=lax` and everything will work.
- Ask users to allow cross site cookies (via browser settings).
- You can use localstorage instead of cookies to store session tokens. However, this has a security issue of exposing tokens via XSS attacks. An example of this can be found [here](https://github.com/supertokens/supertokens-auth-react/tree/master/examples/with-localstorage)

# Browsers where third party cookies are blocked by default:
- Safari
- Incognito mode Google Chrome

**Comment by rishabhpoddar** [comment]
We introduced header based auth which allows use of session tokens without cookies, thereby fixing this issue.
